### PR TITLE
MuSig2 implementation

### PR DESCRIPTION
**Pull request by jimmysong**
Using https://github.com/meshcollider/musig2-py/blob/main/musig2.py as a reference

### CODE REVIEW

**Comment by jimmysong** [comment]
added another change per what I learned from the secp256k1 library review club.

File: buidl/hash.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+try:
```
Comment (mflaxman) [inline buidl/hash.py:1]: I like that this is no longer in `helper.py` (kind of a random catch-call that gets too long).

A little confusing that for ecc ops there's `pecc.py` and `cecc.py` and this uses a different layout. Maybe move the below into `pecc.py` (arguably too long and contains lots of unrelated ECC stuff) or do a similar `phash.py` and `chash.py` file setup?

I don't feel strongly, do whatever you like! Also works as-is. Just feedback on readability.

File: buidl/cecc.py [hunk 2]
```diff
@@ -449,3 +462,19 @@ def parse(cls, wif):
         else:
             raise ValueError("Invalid WIF")
         return cls(secret, network=network, compressed=compressed)
+
+
+def tagged_hash(tag, msg):
```
Comment (mflaxman) [inline buidl/cecc.py:467]: Nice! Curious if this has a meaningful performance impact on the existing (non musig2) stuff.

File: buidl/hash.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    from buidl.cecc import tagged_hash
+except ModuleNotFoundError:
+    import hashlib
+
+    def tagged_hash(tag: bytes, msg: bytes) -> bytes:
+        tag_hash = hashlib.sha256(tag).digest()
+        return hashlib.sha256(tag_hash + tag_hash + msg).digest()
+
+
+def hash_aux(msg):
```
Comment (mflaxman) [inline buidl/hash.py:11]: All of the `tagged_hash`es feels wrong (vs a DRY enum-style construction), but is very readable and using nice helper methods makes it easier for the caller downstream.

Just commentary, probably makes most sense as-is. I get that it was previously this way in `helper.py`, so not really a comment on this PR.

File: buidl/script.py [hunk 4]
```diff
@@ -618,12 +613,17 @@ def address_to_script_pubkey(s):
...
         return P2SHScriptPubKey(h160)
-    elif s[:3] in ("bc1", "tb1"):
+    elif s[:4] in ("bc1q", "tb1q"):
         if len(s) == 42:
             # p2wpkh
             return P2WPKHScriptPubKey(decode_bech32(s)[2])
         elif len(s) == 62:
             # p2wskh
             return P2WSHScriptPubKey(decode_bech32(s)[2])
+    elif s[:4] in ("bc1p", "tb1p"):
```
Comment (mflaxman) [inline buidl/script.py:623]: w00t!

File: buidl/taproot.py [hunk 5]
```diff
@@ -265,15 +271,40 @@ def __init__(self, points, locktime=None, sequence=None):
 
     def get_tweak_point(self, tweak):
         if self.point.parity:
-            return -1 * self.point + tweak
+            tweak_point = -1 * self.point + tweak
         else:
-            return self.point + tweak
+            tweak_point = self.point + tweak
+        return tweak_point
```
Comment (mflaxman) [inline buidl/taproot.py:277]: Nit: no need to define `tweak_point` if all you're doing is returning it (just return it above).

File: buidl/taproot.py [hunk 6]
```diff
@@ -265,15 +271,40 @@ def __init__(self, points, locktime=None, sequence=None):
...
+        )
+        return big_endian_to_int(hash_musignonce(bytes_to_hash))
+
+    def compute_k(self, nonce_secrets, nonce_sums, sig_hash):
+        h = self.compute_coefficient(nonce_sums, sig_hash)
+        return (nonce_secrets[0] + h * nonce_secrets[1]) % N
+
+    def compute_r(self, nonce_sums, sig_hash):
+        h = self.compute_coefficient(nonce_sums, sig_hash)
+        return S256Point.combine([nonce_sums[0], h * nonce_sums[1]])
```
Comment (mflaxman) [inline buidl/taproot.py:301]: The beauty of such well-structure code is how incredibly easy to read the above methods are. Nicely done!

File: buidl/taproot.py [hunk 7]
```diff
@@ -291,7 +322,7 @@ def get_signature(self, s_sum, r, sig_hash, tweak=0):
             msg = r.bip340() + tweak_point.bip340() + sig_hash
             challenge = big_endian_to_int(hash_challenge(msg)) % N
             if tweak_point.parity:
-                s = (s_sum - challenge * tweak) % N
+                s = (-s_sum - challenge * tweak) % N
```
Comment (mflaxman) [inline buidl/taproot.py:325]: What is the reason for this change? Not saying it's wrong but raises a red flag.

File: buidl/test/test_hash.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from unittest import TestCase
```
Comment (mflaxman) [inline buidl/test/test_hash.py:1]: Good catch. Just realized buidl is also missing tests for all the other `tagged_hash` helper methods (i.e. `hash_aux`)

File: buidl/test/test_musig.py [hunk 9]
```diff
@@ -1,17 +1,153 @@
...
+                "e148b34e6d2bb76c2da7efb63b91b69b6bd274aa734a7c837f07a285bce512bb",
+            ],
+        ]
+        for test in tests:
+            want = bytes.fromhex(test[-1])
+            points = [S256Point.parse_bip340(bytes.fromhex(h)) for h in test[:-1]]
+            tap_script = MuSigTapScript(points)
+            self.assertEqual(tap_script.point.bip340(), want)
+
     def test_single_leaf_multisig(self):
```
Comment (mflaxman) [inline buidl/test/test_musig.py:151]: In general better to add a new test-case (without dismantling any of the old one). I think you've eliminated the only test case that calls the `multi_leaf_tap_root` code path?

File: buidl/test/test_musig.py [hunk 10]
```diff
@@ -69,52 +228,86 @@ def test_single_leaf_multisig_locktime(self):
...
+            tx_in.witness.items = []
+            tx_obj.initialize_p2tr_multisig(
+                input_index,
+                tap_root_input.control_block(leaf),
+                tap_root_input.tap_node.tap_script,
+            )
+            sigs = []
+            for priv in private_keys:
+                if priv.point in pubkeys:
+                    sig = tx_obj.get_sig_taproot(input_index, priv, ext_flag=1)
```
Comment (mflaxman) [inline buidl/test/test_musig.py:305]: Named arguments always better for complex methods, especially when it comes to signing logic.

File: buidl/test/test_musig.py [hunk 11]
```diff
@@ -220,50 +475,88 @@ def test_multi_leaf_multisig_locktime(self):
...
+            self.assertTrue(tap_root_input.control_block(leaf))
+            tx_in = tx_ins[input_index]
+            tx_in.witness.items = []
+            tx_obj.initialize_p2tr_multisig(
+                input_index,
+                tap_root_input.control_block(leaf),
+                leaf.tap_script,
+            )
+            sigs = []
+            for priv in private_keys:
```
Comment (mflaxman) [inline buidl/test/test_musig.py:512]: Should there be a helper method for this? These ~8 lines for gathering the sigs are regularly used and quite complex for the downstream caller. Could DRY it out.

File: buidl/test/test_musig.py [hunk 12]
```diff
@@ -277,45 +570,84 @@ def test_multi_leaf_multisig_sequence(self):
...
+                if priv.point in pubkeys:
+                    sig = tx_obj.get_sig_taproot(input_index, priv, ext_flag=1)
+                else:
+                    sig = b""
+                sigs.append(sig)
+            tx_obj.finalize_p2tr_multisig(input_index, sigs)
+            self.assertTrue(tx_obj.verify_input(input_index))
+        input_index = len(tx_ins) - 1
+        tx_in = tx_ins[input_index]
+        musig = MuSigTapScript(points)
```
Comment (mflaxman) [inline buidl/test/test_musig.py:613]: Ditto the below, should this have a helper method? Seems not only very advanced for the caller but something where you want to centralize the logic in a place that has nice test coverage.

File: buidl/block.py [hunk 13]
```diff
@@ -177,3 +179,28 @@ def get_outpoints(self):
...
+        )
+    ),
+    "signet": Block.parse_header(
+        BytesIO(
+            bytes.fromhex(
+                "0100000000000000000000000000000000000000000000000000000000000000000000003ba3edfd7a7b12b27ac72c3e67768f617fc81bc3888a51323a9fb8aa4b1e5e4a008f4d5fae77031e8ad22203"
+            )
+        )
+    ),
+}
```
[the remark this answers is not in the document]
Comment (mflaxman) [reply buidl/block.py:206]: Also, could be a use-case for allowing `parse_header()` to accept hex in some form:
* `def parse_header_from_hex(hex):`
* `def parse_header(s, hex=""):`

**Review by mflaxman — APPROVED** [review]
Sorry for the delay, busy with travel + catching up from travel.

This is a *really* impressive PR, nicely done! Definitely bump that version number in `setup.py`, this warrants a new release.

Want to add it to `multiwallet.py`? Would be awesome to have this available for people to use in the wild (you could default it to testnet with a big scary disclaimer if you like).

File: buidl/cecc.py (hunk 2 above)
Comment (jimmysong) [reply buidl/cecc.py:467]: I haven't checked, but thought it would be better to be consistent. I know it'll make an impact once you get into 10k hashes or so.

File: buidl/taproot.py (hunk 5 above)
Comment (jimmysong) [reply buidl/taproot.py:277]: right, I forgot to change this back when I was debugging.

File: buidl/taproot.py (hunk 7 above)
Comment (jimmysong) [reply buidl/taproot.py:325]: Turns out there was a flaw in my logic that didn't get revealed until this line got tested (which didn't happen until the new musig2 tests)! I ran through some math and this is the correct formula.

File: buidl/test/test_musig.py [hunk 14]
```diff
@@ -1,17 +1,153 @@
...
 from buidl.hd import HDPrivateKey
 from buidl.helper import SIGHASH_DEFAULT
+from buidl.script import address_to_script_pubkey
 from buidl.taproot import MultiSigTapScript, MuSigTapScript, TapRootMultiSig
 from buidl.timelock import Locktime, Sequence
 from buidl.tx import Tx, TxIn, TxOut
 from buidl.witness import Witness
 
 
 class MuSigTest(TestCase):
```
[the remark this answers is not in the document]
Comment (jimmysong) [reply buidl/test/test_musig.py:14]: Yea, I generated these using meshcollider's MuSig2 python implementation. He uses random points, so I can't regenerate them, but that's what his library said.

File: buidl/test/test_musig.py (hunk 9 above)
Comment (jimmysong) [reply buidl/test/test_musig.py:151]: `test_multi_leaf_multisig` uses `multi_leaf_tap_root`

File: buidl/test/test_musig.py (hunk 11 above)
Comment (jimmysong) [reply buidl/test/test_musig.py:512]: That's basically what `finalize_p2tr_multisig` does. The loop is there for the coordinator to gather the `sigs` array. In practice, the stuff in the for loop will most likely be done by different keys and generated so this is really a way to simulate the signing.

I could DRY out this since it gets repeated, but I've found doing that in tests tends to not be a good idea as changing the helper test function tends to break lots of tests at once.

File: buidl/test/test_musig.py (hunk 12 above)
Comment (jimmysong) [reply buidl/test/test_musig.py:613]: We're simulating each key signer generating two nonces for the MuSig2 protocol. This is where you have multiple signers produce their own nonces so it doesn't make much sense to have a function that generates all of them at once as that shouldn't be happening. I'm also against a helper function within the test suite because that tends to make the tests less readable and one change tends to break lots of tests. Also, those functions tend to get very complicated as just one more condition to add yet another test case makes it really hard to debug. I generally just like having test cases be really readable and not call other stuff that's not in the library.

**Comment by mflaxman** [comment]
Woooooooohoooooo!

@jimmysong any plans to implement this using multiwallet CLI? It would be *really* cool to have a usable musig2 implementation live that people can play with (could easily limit/default to testnet). Happy to help with the UX if that's a pain holding you back.

**Comment by jimmysong** [comment]
sure. let me do that so there's something you can play with. then I'm going to see what I can do with FROST. Our friend S. Snigirev will be giving a talk on it at the conference I'm attending next week.

**Comment by mflaxman** [comment]
Awesome, very exciting!

buidl is now bleeding edge :)
